components/affix/affix.directive.ts

```diff
@@ -0,0 +1,174 @@
+    private isAffixedBottom:boolean = true;
+    @HostBinding('style.top.px')
+    private top:number = null;
+
+    @Output()
+    public affixChange:EventEmitter<AffixStatusChange> = new EventEmitter(false);
+
+    private status:AffixStatus = null;
+    private body:HTMLBodyElement;
+    private window:Window;
+    private pinnedOffset:number = null;
+    private debouncedCheckPosition:Function = Affix.debounce(() => this.checkPosition(), 5);
+    private eventListener:Function = (ev:UIEvent) => this.debouncedCheckPosition();
+
+    constructor(private el:ElementRef) {
+        this.body = el.nativeElement.ownerDocument.body;
+        this.window = el.nativeElement.ownerDocument.defaultView;
+    }
+
+    ngOnInit() {
+        this.el.nativeElement.ownerDocument.defaultView.addEventListener('scroll', this.eventListener);
+        this.checkPosition();
+    }
+
+    ngOnDestroy():any {
+        this.el.nativeElement.ownerDocument.defaultView.removeEventListener('scroll', this.eventListener);
+        return undefined;
+    }
+
+    private checkPosition():void {
+        let elemPos = positionService.position(this.el.nativeElement);
+        if (elemPos.height === 0 || elemPos.width === 0) {
+            // Element is not visible
+            return;
+        }
+        let scrollHeight:number = Math.max(this.window.innerHeight, this.body.scrollHeight);
+        let nativeElemPos:ElemPosition = positionService.offset(this.el.nativeElement);
+
+        let newAffixStatus:AffixStatus = this.getState(scrollHeight, nativeElemPos, this.affixOffsetTop, this.affixOffsetBottom);
+
+        if (this.status !== newAffixStatus) {
+
+            this.top = newAffixStatus === AffixStatus.AFFIX_BOTTOM ? this.getPinnedOffset() : null;
+
+            this.affixChange.emit(new AffixStatusChange(this.status, newAffixStatus));
+            this.status = newAffixStatus;
+            this.isAffix = false;
+            this.isAffixedBottom = false;
+            this.isAffixedTop = false;
+            switch (this.status) {
+                case AffixStatus.AFFIX_TOP:
+                    this.isAffixedTop = true;
+                    break;
+                case AffixStatus.AFFIX_BOTTOM:
+                    this.isAffixedBottom = true;
+                    break;
+                default:
+                    this.isAffix = true;
+                    break;
+            }
+        }
+
+        if (newAffixStatus === AffixStatus.AFFIX_BOTTOM) {
+            this.top = scrollHeight - nativeElemPos.height - this.affixOffsetBottom;
+        }
+    }
+
+    private getState(scrollHeight:number, nativeElemPos:ElemPosition, offsetTop:number, offsetBottom:number):AffixStatus {
+        let scrollTop:number = this.body.scrollTop; // current scroll position in pixels from top
+        let targetHeight:number = this.window.innerHeight; // Height of the window / viewport area
+
+        if (offsetTop !== null && this.status === AffixStatus.AFFIX_TOP) {
```

Comment:
Thanks for the explanation.

Another thing I noticed when testing in Firefox is that `body.scrollTop` is always 0.  Digging around, I found https://stackoverflow.com/questions/28633221/document-body-scrolltop-firefox-returns-0-only-js, which suggests that you might need to expand your approach for calculating `scrollTop` (here and in `getPinnedOffset()`).